ante/fee_utils.go

```diff
@@ -0,0 +1,190 @@
+//
+// overwrite DenomsSubsetOfIncludingZero from sdk, to allow zero amt coins in superset. e.g. 1stake is DenomsSubsetOfIncludingZero 0stake. [] is the DenomsSubsetOfIncludingZero of [0stake] but not [1stake].
+// DenomsSubsetOfIncludingZero returns true if coins's denoms is subset of coinsB's denoms.
+// if coins is empty set, empty set is any sets' subset
+func DenomsSubsetOfIncludingZero(coins, coinsB sdk.Coins) bool {
+	// more denoms in B than in receiver
+	if len(coins) > len(coinsB) {
+		return false
+	}
+	// coins=[], coinsB=[0stake]
+	// todo let all len(coins) == 0 pass and reject later at IsAnyGTEIncludingZero ?
+	if len(coins) == 0 && containZeroCoins(coinsB) {
+		return true
+	}
+	// coins=1stake, coinsB=[0stake,1uatom]
+	for _, coin := range coins {
+		err := sdk.ValidateDenom(coin.Denom)
+		if err != nil {
+			panic(err)
+		}
+
+		if ok, _ := coinsB.Find(coin.Denom); !ok {
+			return false
+		}
+	}
+
+	return true
+}
+
+// overwrite the IsAnyGTEIncludingZero from sdk to allow zero coins in coins and coinsB.
+// IsAnyGTEIncludingZero returns true if coins contain at least one denom that is present at a greater or equal amount in coinsB; it returns false otherwise.
+// if CoinsB is emptyset, no coins sets are IsAnyGTEIncludingZero coinsB unless coins is also empty set.
+// NOTE: IsAnyGTEIncludingZero operates under the invariant that both coin sets are sorted by denoms.
+// contract !!!! coins must be DenomsSubsetOfIncludingZero of coinsB
+func IsAnyGTEIncludingZero(coins, coinsB sdk.Coins) bool {
+	// no set is empty set's subset except empty set
+	// this is different from sdk, sdk return false for coinsB empty
+	if len(coinsB) == 0 && len(coins) == 0 {
+		return true
+	}
+	// nothing is gte empty coins
+	if len(coinsB) == 0 && len(coins) != 0 {
+		return false
+	}
+	// if feecoins empty (len(coins)==0 && len(coinsB) != 0 ), and globalfee has one denom of amt zero, return true
+	if len(coins) == 0 {
+		return containZeroCoins(coinsB)
+	}
+
+	//  len(coinsB) != 0 && len(coins) != 0
+	// special case: coins=1stake, coinsB=[2stake,0uatom], fail
+	for _, coin := range coins {
+		// not find coin in CoinsB
+		if ok, _ := coinsB.Find(coin.Denom); ok {
+			// find coin in coinsB, and if the amt == 0, mean either coin=0denom or coinsB=0denom...both true
+			amt := coinsB.AmountOf(coin.Denom)
+			if coin.Amount.GTE(amt) {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+// return true if coinsB is empty or contains zero coins,
+// CoinsB must be validate coins !!!
+func containZeroCoins(coinsB sdk.Coins) bool {
+	if len(coinsB) == 0 {
+		return true
+	}
+	for _, coin := range coinsB {
+		if coin.IsZero() {
+			return true
+		}
+	}
+
+	return false
+}
+
+// CombinedFeeRequirement will combine the global fee and min_gas_price. globalFees, minGasPrices must be valid, but CombinedFeeRequirement does not valid them so they may return 0denom.
```

Comment:
```suggestion
// CombinedFeeRequirement will combine the global fee and min_gas_price. Both globalFees and minGasPrices must be valid, but CombinedFeeRequirement does not validate them so it may return 0denom.
```